client/files.go

```diff
@@ -0,0 +1,134 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+)
+
+// MakeDirOptions holds the options for a call to MakeDir.
+type MakeDirOptions struct {
+	// Path is the absolute path of the directory to be created (required).
+	Path string
+
+	// MakeParents, if true, specifies that any non-existent parent directories
+	// should be created. If false (the default), the call will fail if the
+	// directory to be created has at least one parent directory that does not
+	// exist.
+	MakeParents bool
+
+	// Permissions specifies the permission bits of the directories to be created.
+	// If 0 or unset, defaults to 755.
+	Permissions os.FileMode
+
+	// UserID indicates the ID of the owner user for the created directories.
+	UserID *int
+
+	// User indicates the name of the owner user for the created directories.
+	// If used together with UserID, this value must match the name of the user
+	// with that ID.
+	User string
+
+	// GroupID indicates the ID of the owner group for the created directories.
+	GroupID *int
+
+	// Group indicates the name of the owner group for the created directories.
+	// If used together with GroupID, this value must match the name of the group
+	// with that ID.
+	Group string
+}
+
+type makeDirPayload struct {
+	Action string         `json:"action"`
+	Dirs   []makeDirsItem `json:"dirs"`
+}
+
+type makeDirsItem struct {
+	Path        string `json:"path"`
+	MakeParents bool   `json:"make-parents"`
+	Permissions string `json:"permissions"`
+	UserID      *int   `json:"user-id"`
+	User        string `json:"user"`
+	GroupID     *int   `json:"group-id"`
+	Group       string `json:"group"`
+}
+
+type fileResult struct {
+	Path  string       `json:"path"`
+	Error *errorResult `json:"error,omitempty"`
+}
+
+type errorResult struct {
+	Message string      `json:"message"`
+	Kind    string      `json:"kind,omitempty"`
+	Value   interface{} `json:"value,omitempty"`
+}
+
+// MakeDir creates a directory or directory tree.
+// The error returned is a *Error if the request went through successfully
+// but there was an OS-level error creating the directory, with the Kind
+// field set to the specific error kind, for example "permission-denied".
+func (client *Client) MakeDir(opts *MakeDirOptions) error {
+	var permissions string
+	if opts.Permissions != 0 {
+		permissions = fmt.Sprintf("%03o", opts.Permissions)
```

Comment:
I was not sure why we are printing this in at least a 3-rune form. It looks like the daemon tests for either zero or 3 rune length string when validating permissions.

Note that treating the directory permissions as only 3 octal values may not be correct. At least chisel recently added support for the sticky bit if I read this correctly. 

https://github.com/canonical/chisel/commit/c06b76261f1399256655dddc99196c28cb592291

This means a change in the daemon handler.

Also see: https://github.com/golang/go/issues/25539

Reply:
This is the common logic I've found across pebble, but I was secretly thinking on unifying that behavior once all the file-related PRs are merged, perhaps by adding this utility function to x-go:

https://github.com/canonical/pebble/blob/b13464b97acd35b140b1ad2758ce745f0735daf8/internal/daemon/api_files.go#L514-L523

And, of course, treat the permissions properly there. So, I'm not sure if I should directly fix it here, implement a second `parsePermissions()` function on files.go with the correct behavior, or leave it like this and fix it once these PRs are merged.

What do you think?

Reply:
I guess breaking it up makes sense.